### PR TITLE
Rework type

### PR DESCRIPTION
**Pull request by ph-fritsche**
**What**:

This PR will include a complete rewrite of `userEvent.type` to tackle the issues tracked in #521 

There are a lot of changes and some of the fixes along the way might break code that considered these bugs features. So a major version update should happen.

With the changes `userEvent.type` will actually return no Promise if there is no `delay`.

The newly introduces `userEvent.keyboard` skips all convenience of `userEvent.type` and just applies the key events on the active element.
It allows to apply events and continue with a keyboard state.

https://github.com/testing-library/user-event/blob/b8a78c8f9dc9d19b0ef57ffdeef8e4af8d62f3a8/src/__tests__/keyboard/index.ts#L111-L137

**Why**:

`userEvent.type` currently is a patchwork of specialized implementations for edge cases.
We've got a bunch of open issues that nobody wants to tackle because it is unclear where exactly a change should be implemented and if there are more things to consider.
The code lacks structure and documentation.

The conversion to Typescript highlighted a lot of hidden debt inside the code regarding assumptions about return values passed around without further checks.

**How**:

- [x] First Step: Split `utils` and `type` into multiple files and convert to Typescript to improve confidence in upcoming changes
- [x] Refactor the `typeImplementation` to provide a shared internal API to extend for special characters, meta characters, etc
- [x] Convert handler implementations to new plugins
- [x] Add tests for uncovered lines
- [x] Introduce new API
- [x] Add documentation

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Typings
- [x] Ready to be merged

### CODE REVIEW

**Comment by ph-fritsche** [comment]
@adamayres The rewrite is complete. All previous tests (apart from a) a bug and b) an unlogic edge case) pass.
Your bug should be fixed by the new implementation too.

**Comment by eps1lon** [comment]
I'm not familiar with `user-event` and don't intend to use. I wouldn't have the time to review this change.

File: README.md [hunk 1]
```diff
@@ -271,6 +268,76 @@ test('types into the input', () => {
...
+  ```
+  This does not keep any key pressed. So `Shift` will be lifted before pressing
+  `f`.
+- Per
+  [KeyboardEvent.code](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code)
+  ```js
+  userEvent.keyboard('[ShiftLeft][KeyF][KeyO][KeyO]') // translates to: Shift, f, o, o
+  ```
+- Per legacy `userEvent.type` modifier/specialChar The modifiers like `{shift}`
+  (note the lowercase) will automatically be kept pressed, just like before. You
```
Comment (kentcdodds) [inline README.md:305]: Interesting that there's inconsistency in whether it's upper or lower case. Any chance we could make it just not matter and it can be cased however folks want?

File: README.md [hunk 2]
```diff
@@ -271,6 +268,76 @@ test('types into the input', () => {
...
+  userEvent.keyboard('{Shift}{f}{o}{o}') // translates to: Shift, f, o, o
+  ```
+  This does not keep any key pressed. So `Shift` will be lifted before pressing
+  `f`.
+- Per
+  [KeyboardEvent.code](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code)
+  ```js
+  userEvent.keyboard('[ShiftLeft][KeyF][KeyO][KeyO]') // translates to: Shift, f, o, o
+  ```
+- Per legacy `userEvent.type` modifier/specialChar The modifiers like `{shift}`
```
Comment (kentcdodds) [inline README.md:304]: Calling it "legacy" makes it sound like we're recommending *against* using `userEvent.type`. If I understand these changes correctly, that's not necessarily the case. Could do:

```suggestion
- Per `userEvent.type`-like modifier/specialChar modifiers like `{shift}`
```

File: src/__tests__/keyboard/getNextKeyDef.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
+import {getNextKeyDef} from 'keyboard/getNextKeyDef'
```
Comment (kentcdodds) [inline src/__tests__/keyboard/getNextKeyDef.ts:1]: Just an idea, not necessarily a recommendation. I think this could be a good place for `jest-in-case` as lots of these tests look a lot like each other.

File: src/__tests__/keyboard/index.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    body - keyup: o (111)
+    body - keydown: o (111)
+    body - keypress: o (111)
+    body - keyup: o (111)
+  `)
+})
+
+it('type with focus', () => {
+  const {element} = setup('<input/>')
+  const {getEventSnapshot} = addListeners(document.body)
```
Comment (kentcdodds) [inline src/__tests__/keyboard/index.ts:28]: Curious why this isn't:

```suggestion
  const {element, getEventSnapshot} = setup('<input/>')
```

File: src/__tests__/type.js [hunk 5]
```diff
@@ -792,9 +792,7 @@ test('typing an invalid input value', () => {
   const {element} = setup('<input type="number" />')
   userEvent.type(element, '3-3')
 
-  // TODO: fix this bug
-  // THIS IS A BUG! It should be expect(element.value).toBe('')
-  expect(element).toHaveValue(-3)
+  expect(element).toHaveValue(null)
```
Comment (kentcdodds) [inline src/__tests__/type.js:795]: HOORAY! A TODO that's actually been addressed. It's a miracle 😆

File: src/keyboard/getEventProps.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+export function getKeyEventProps(keyDef: keyboardKey, state: keyboardState) {
+  return {
+    key: keyDef.key,
+    code: keyDef.code,
+    altKey: state.modifiers.alt,
+    ctrlKey: state.modifiers.ctrl,
+    metaKey: state.modifiers.meta,
+    shiftKey: state.modifiers.shift,
+
+    // deprecated
```
Comment (kentcdodds) [inline src/keyboard/getEventProps.ts:12]: If we use the `@deprecated` annotation then vscode/TypeScript will offer some suggestions :)

```suggestion
    /** @deprecated use code instead */
```

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAcwKZQNKoJ4FEBuqYUACgE5wAOAzgBQCUiA3gLABQiiZ6IZSrHTogDWOAFyIA5KOwARVMAB0MyQBp2QxBDgATVBOk55S7XrUahAQwA2mcVOpRLUVIoC2umMBioy1RTZQ5oKc0GTWWNgGjs6uHjpePn6KYdbBmm7olpHRTi7unt6+-plO6ULUABZedlEOeXGFSf5VNcEWnAD0AFTdiAACepTcELE6iCDUqFq60zBgjqiW492dHSI4AMKzBtWSFgC+7Eds7GenbNoLUFyo1IgAvCjokQREpBQ0DOfc-jLbeiAA

image

File: src/type/typeImplementation.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
+// TODO: wrap in asyncWrapper
```
Comment (kentcdodds) [inline src/type/typeImplementation.ts:1]: I think this is accomplished in `src/type/index.ts` right?

```suggestion
```

File: src/type/typeImplementation.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+      initialSelectionEnd ?? value.length,
+    )
+  }
+
+  // previous implementation did wait before the first character
+  // this probably can be removed
+  if (delay > 0) {
+    await wait(delay)
+  }
+
```
Comment (kentcdodds) [inline src/type/typeImplementation.ts:76]: Agreed, if we're breaking things anyway we may as well remove this which doesn't have much reason to exist.

```suggestion
```

**Review by kentcdodds — APPROVED** [review]
Looking great! Fantastic work 👏👏👏

A few general comments:

1. I typically avoid camelCasing filenames due to inconsistencies in operating systems with filename casing. I use kebab-case instead
2. There are a lot of tiny files. I've done things that way in the past and I've found that things are simpler when you only split things out when necessary. I'm mostly thinking about the utils/misc stuff.

In any case, you're the primary maintainer at this point, so you can think of my comments/questions as ideas and suggestions and you're more than welcome to ignore any and all of them :)

With that said, I'll give you my green stamp of approval because I think what you have here is awesome. Thank you for all you do!

File: README.md [hunk 9]
```diff
@@ -271,6 +268,76 @@ test('types into the input', () => {
...
+Keystrokes can be described:
+
+- Per printable character
+  ```js
+  userEvent.keyboard('foo') // translates to: f, o, o
+  ```
+  The brackets `{` and `[` are used as special character and can be referenced
+  by doubling them.
+  ```js
+  userEvent.keyboard('{{a[[') // translates to: {, a, [
```
[the remark this answers is not in the document]
Comment (ph-fritsche) [reply README.md:289]: I'm not sure what I prefer. Escaping in strings is always nasty. I just went forward interpreting the string as when encountering `{` or `[` treat the following string in a special way. A little bit like some template languages like Twig.

File: README.md [hunk 10]
```diff
@@ -271,6 +268,76 @@ test('types into the input', () => {
...
+- Per
+  [KeyboardEvent.code](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code)
+  ```js
+  userEvent.keyboard('[ShiftLeft][KeyF][KeyO][KeyO]') // translates to: Shift, f, o, o
+  ```
+- Per legacy `userEvent.type` modifier/specialChar The modifiers like `{shift}`
+  (note the lowercase) will automatically be kept pressed, just like before. You
+  can cancel this behavior by adding a `/` to the end of the descriptor.
+  ```js
+  userEvent.keyboard('{shift}{ctrl/}a{shift}') // translates to: Shift(down), Control(down+up), a, Shift(up)
```
[the remark this answers is not in the document]
Comment (ph-fritsche) [reply README.md:308]: Good catch. Thanks

File: README.md (hunk 1 above)
Comment (ph-fritsche) [reply README.md:305]: Sadly not. As the previous behavior was different for different keys.
The new implementation is consistent for all keys and adds an exception for the old modifiers.

File: README.md (hunk 2 above)
Comment (ph-fritsche) [reply README.md:304]: No, but I would recommend against the old modifiers as they behave differently for different keys.

File: src/type/typeImplementation.ts (hunk 7 above)
Comment (ph-fritsche) [reply src/type/typeImplementation.ts:1]: I was not sure what this TODO was referring to. Might have been accomplished a long time ago.

File: src/__tests__/keyboard/index.ts (hunk 4 above)
Comment (ph-fritsche) [reply src/__tests__/keyboard/index.ts:28]: Because `keyboard` does not move the focus.
If the focus is not on the `input`, there are no events on the input.
The eventSnapshot on `document.body` tells the full story for these tests and provides a meaningful error message both for wrong events and wrong event targets.

File: src/__tests__/keyboard/getNextKeyDef.ts (hunk 3 above)
Comment (ph-fritsche) [reply src/__tests__/keyboard/getNextKeyDef.ts:2]: I implemented it. Not sure if I like the look of the test file. But it did get rid of some repetition. :)

**Comment by ph-fritsche** [comment]
> 2. There are a lot of tiny files. I've done things that way in the past and I've found that things are simpler when you only split things out when necessary. I'm mostly thinking about the utils/misc stuff.

Yes, that might have been overkill. But I went crazy jumping through code when figuring out what interacts / is used where and what to reuse/rewrite/remove. So I just went ahead and split everything up before doing any changes.
I think when restructuring code it's a lot easier to deal with a bunch of files than multiple functions in different positions in a shared file with others you don't want to touch at that moment. But it might be just personal preference.

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (kentcdodds) [reply README.md:289]: Sounds fine to me. I agree escaping in strings is really annoying and using `\` as the mechanism just makes it more challenging. This is probably fine +

File: README.md (hunk 2 above)
Comment (kentcdodds) [reply README.md:304]: Makes sense 👍

File: src/__tests__/keyboard/index.ts (hunk 4 above)
Comment (kentcdodds) [reply src/__tests__/keyboard/index.ts:28]: Ah, yeah, that makes sense 👍

**Review by kentcdodds — APPROVED** [review]
Thanks for answering my questions and for all your hard work on this project. It's fantastic 👍 

You've got a big green ✅ from me!
